### PR TITLE
Feat / 94 toggle button

### LINKED ISSUE #94: Toolkit: toggle button
- [ ] design toggle button

### PR DESCRIPTION
**Pull request by sam-keen**
Closes #94

### CODE REVIEW

File: libs/ui-toolkit/src/components/toggle/toggle.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        },
+      ]}
+      checkedValue={checked}
+      onChange={(e) => setChecked(e.target.value)}
+      className="max-w-[400px]"
+    />
+  );
+};
+
+// @ts-ignore args provided after
```
Comment (dexturr) [inline libs/ui-toolkit/src/components/toggle/toggle.stories.tsx:33]: chore (blocking): I think but explicitly declaring the args type to be `ToggleInputProps` this should be removable

Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.stories.tsx:33]: Oh nice, will try that, thanks

File: libs/ui-toolkit/src/components/toggle/toggle.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+export default {
+  component: Toggle,
+  title: 'Toggle',
+} as ComponentMeta<typeof Toggle>;
+
+export const Controlled: ComponentStory<typeof Toggle> = () => {
+  const [checked, setChecked] = useState('test-1');
+
+  return (
+    // Wrapping in a form to scope the effect of the radios (stops light and dark theme versions colliding)
```
Comment (dexturr) [inline libs/ui-toolkit/src/components/toggle/toggle.stories.tsx:14]: nitpick (non-blocking): all controls should be in forms anyway

```suggestion
```

File: libs/ui-toolkit/src/components/toggle/toggle.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+export const Toggle = ({
+  name,
+  toggles,
+  className,
+  onChange,
+  checkedValue,
+  ...props
+}: ToggleInputProps) => {
+  return (
+    <fieldset className={`${className} flex`} {...props}>
```
Comment (dexturr) [inline libs/ui-toolkit/src/components/toggle/toggle.tsx:32]: chore (non-blocking): can we use the classNames helper here?

**Review by dexturr — CHANGES_REQUESTED** [review]
Be good to bulk out the tests to the functionality as well as the dom

File: libs/ui-toolkit/src/components/toggle/toggle.stories.tsx (hunk 2 above)
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.stories.tsx:14]: Aye, it's a good nitpick!

File: libs/ui-toolkit/src/components/toggle/toggle.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+export default {
+  component: Toggle,
+  title: 'Toggle',
+} as ComponentMeta<typeof Toggle>;
+
+export const Controlled: ComponentStory<typeof Toggle> = () => {
+  const [checked, setChecked] = useState('test-1');
+
+  return (
+    <form>
```
Comment (mattrussell36) [inline libs/ui-toolkit/src/components/toggle/toggle.stories.tsx:14]: suggestion (non blocking): Might be nice to render the current selected state here to demonstrate controlled mode.

**Review by mattrussell36 — CHANGES_REQUESTED** [review]
Looks good, I think we might as well utilise the Toggle component immediately as part of this PR. Could you add it to the deal ticket and to the chart (I think those are the only two places its used)

File: libs/ui-toolkit/src/components/toggle/toggle.tsx (hunk 3 above)
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.tsx:32]: I wouldn't choose to myself here as I don't think it makes things any simpler, but if it creates consistency with other areas, sure

File: libs/ui-toolkit/src/components/toggle/toggle.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { render, screen } from '@testing-library/react';
+
+import { Toggle } from './toggle';
+
+describe('Toggle', () => {
+  it('should render button successfully', () => {
```
[the remark this answers is not in the document]
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.spec.tsx]: Note: the way I set up the component en
Screenshot 2022-04-12 at 10 40 50
sures the toggles array cannot be empty

File: libs/ui-toolkit/src/components/toggle/toggle.stories.tsx (hunk 4 above)
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.stories.tsx:14]: Good idea, thanks

File: libs/ui-toolkit/src/components/toggle/toggle.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import type { ChangeEvent } from 'react';
+
+// Supports controlled and uncontrolled setups.
+interface ToggleProps {
+  label: string;
+  value: string;
+}
+
+type TogglesArr = [
```
[the remark this answers is not in the document]
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.tsx:9]: It does indeed!

Screenshot 2022-04-12 at 10 40 50

File: libs/ui-toolkit/src/index.ts [hunk 7]
```diff
@@ -26,3 +27,6 @@ export { Card } from './components/card';
 
 // Utils
 export * from './utils/intent';
+
+// Types
+export type { TogglesArr } from './components/toggle';
```
Comment (dexturr) [inline libs/ui-toolkit/src/index.ts:32]: question: why needed?

File: libs/ui-toolkit/src/components/toggle/toggle.tsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
+import classnames from 'classnames';
+import type { ChangeEvent } from 'react';
+
+// Supports controlled and uncontrolled setups.
+interface ToggleProps {
+  label: string;
+  value: string;
+}
+
+export type TogglesArr = [ToggleProps, ToggleProps, ...ToggleProps[]];
```
Comment (dexturr) [inline libs/ui-toolkit/src/components/toggle/toggle.tsx:10]:
```suggestion
```

File: libs/ui-toolkit/src/components/toggle/toggle.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+export interface ToggleInputProps {
+  name: string;
+  toggles: TogglesArr;
```
Comment (dexturr) [inline libs/ui-toolkit/src/components/toggle/toggle.tsx:14]:
```suggestion
  toggles: ToggleProps[];
```

File: libs/ui-toolkit/src/components/toggle/toggle.tsx (hunk 8 above)
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.tsx:10]: This ensures it expects a minimum of two args

File: libs/ui-toolkit/src/components/toggle/toggle.tsx (hunk 9 above)
Comment (sam-keen) [reply libs/ui-toolkit/src/components/toggle/toggle.tsx:14]: We lose TS check for min 2 args doing this

**Review by dexturr — APPROVED** [review]

**Review by dexturr — APPROVED** [review]

**Review by mattrussell36 — APPROVED** [review]

File: libs/deal-ticket/src/type-selector.tsx [hunk 10]
```diff
@@ -9,16 +10,18 @@ interface TypeSelectorProps {
...
         name="order-type"
-        options={Object.entries(OrderType).map(([text, value]) => ({
-          text,
-          value,
-        }))}
-        currentOption={order.type}
-        onSelect={(value) => onSelect(value as OrderType)}
+        toggles={toggles as TogglesArr}
+        checkedValue={order.type}
+        onChange={(e) => onSelect(e.target.value as OrderType)}
```
[the remark this answers is not in the document]
Comment (sam-keen) [reply libs/deal-ticket/src/type-selector.tsx:23]: Hi @mattrussell36, I can't think of a way that doesn't either

a. make the toggle less reusable
b. end up with more complexity in what we need to pass to the toggle

As this is non-blocking I'm going to merge this, but happy to discuss further! Cheers

**Pull request merged**